File: modules/flutter_channel/android/src/main/java/com/example/flutter_channel/FlutterBridge.java
package com.example.flutter_channel;

import android.os.Handler;
import android.os.Looper;

import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.flutter.plugin.common.BasicMessageChannel;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.StandardMessageCodec;


public class FlutterBridge implements BasicMessageChannel.MessageHandler<Map<String, Object>> {

   class MainThreadRunnable {

      private Runnable runnable;
      private Handler handler;

      MainThreadRunnable(Runnable runnable) {
         this.runnable = runnable;
         handler = new Handler(Looper.getMainLooper());

         handler.post(this.runnable);
      }
   }

   //静态内部类
   private static class FlutterBridgeHolder {
      private static final FlutterBridge sInstance = new FlutterBridge();
   }

   private FlutterBridge() {
   }

   public static FlutterBridge getInstance() {
      //第一次调用getInstance方法时才加载FlutterBridgeHolder并初始化sInstance
      return FlutterBridgeHolder.sInstance;
   }

   private BinaryMessenger binaryMessenger;

   private FlutterBridgeDelegate _delegate;

   private BasicMessageChannel _nativeChannel;

   public static void setupFlutterBinaryMessenger(BinaryMessenger messenager) {
      FlutterBridge.getInstance().binaryMessenger = messenager;
      FlutterBridge.getInstance().setNativeMessageHandler();
   }

   /// 利用自定义配置进行初始化
   /// @param delegate FlutterBridgeDelegate的实例，用于实现methodName 的处理方法
   public void setupDelegate(FlutterBridgeDelegate delegate) {
      this._delegate = delegate;
   }

   void setNativeMessageHandler() {
      if (_nativeChannel == null) {
         _nativeChannel = new BasicMessageChannel<>(this.binaryMessenger, "flutter.sendEvent.native", new StandardMessageCodec());

         _nativeChannel.setMessageHandler(this);
      }
   }

   @Override
   public void onMessage(@Nullable Map<String, Object> message, @NonNull BasicMessageChannel.Reply<Map<String, Object>> reply) {
      if (message != null) {

         if (_delegate == null) {
            reply.reply(_getErrorMap(-3));
            return;
         }

         CallMethodParams params = CallMethodParams.fromMap(message);

         if (params.getMethodName() == null || params.getMethodName().length() == 0) {
            reply.reply(_getErrorMap(-2));
            return;
         }

         String methodName = params.getMethodName();

         _delegate.callHandler(methodName, params.getArguments(), (value) -> {
            if (value != null) {
               sendCallbackToFlutter(methodName, value);
            }
         });
         reply.reply(new HashMap<>());
         return;
      }
      reply.reply(_getErrorMap(-1));
   }


   Map<String, Object> _getErrorMap(int code) {
      HashMap<String, String> map = new HashMap<>();

      if (code == -1) {
         map.put("code", "-1001");
         map.put("message", "message 格式不正确");
         map.put("details", "message 格式不正确!!!");
      } else if (code == -2) {
         map.put("code", "-1002");
         map.put("message", "methodName 不可为空");
         map.put("details", "methodName 不可为空!!!");
      } else if (code == -3) {
         map.put("code", "-1003");
         map.put("message", "delegate 未设置");
         map.put("details", "delegate 未设置，请使用setupDelegate初始化代理方法");
      }

      HashMap<String, Object> _t_map = new HashMap<>();
      _t_map.put("error", map);
      return _t_map;
   }

   /// 将自定义事件传递给flutter侧
   /// @param key 事件的标识符
   /// @param arguments 事件的参数
   public void sendEventToFlutter(String key, Map<String, Object> arguments) {
      CommonParams params = new CommonParams(key);
      params.arguments = arguments;

      BasicMessageChannel channel = new BasicMessageChannel<>(binaryMessenger, "flutter.sendEvent.listener", new StandardMessageCodec());

      new MainThreadRunnable(() -> {
         channel.send(params.toMap(), (reply) -> {

         });
      });
   }

   /// 将自定义事件传递给flutter侧
   /// @param methodName 调用原生端的方法名称
   /// @param arguments 事件的参数
   public void sendCallbackToFlutter(String methodName, Map<String, Object> arguments) {
      ResponseParams params = new ResponseParams(methodName);
      params.arguments = arguments;

      new MainThreadRunnable(() -> {
         _nativeChannel.send(params.toMap(), (reply) -> {

         });
      });
   }

}

File: modules/flutter_channel/ios/Classes/FlutterBridge.m
//
//  FlutterBridge.m
//  flutter_channel
//
//  Created by 赵光飞 on 2022/1/27.
//

#import "FlutterBridge.h"
#import <Flutter/Flutter.h>
#import "Messages.h"
#import "FlutterBridgeDelegate.h"
#import "NativeViewDelegate.h"

@interface FlutterBridge ()

@property (nonatomic, strong) NSObject<FlutterBinaryMessenger> *binaryMessenger;

@property (nonatomic, strong) id<FlutterBridgeDelegate> delegate;

@property (nonatomic, strong, readwrite) id<NativeViewDelegate> viewDelegate;

@property (nonatomic, strong) FlutterBasicMessageChannel *nativeChannel;

@end

@implementation FlutterBridge

+ (instancetype)instance{
    static id _instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [self.class new];
    });
    
    return _instance;
}

+ (void)setupFlutterBinaryMessenger:(NSObject<FlutterBinaryMessenger> *)messenager {
    [FlutterBridge instance].binaryMessenger = messenager;
    [[FlutterBridge instance] setNativeMessageHandler];
}

- (void)setupDelegate:(id<FlutterBridgeDelegate>)delegate {
    _delegate = delegate;
}

- (void)setupViewDelegate:(id<NativeViewDelegate>)viewDelegate {
    _viewDelegate = viewDelegate;
}

- (void)sendEventToFlutterWith:(NSString *)key arguments:(NSDictionary *)arguments {
    [self sendEventToFlutterWith:key arguments:arguments completion:^(NSError * _Nullable error) {
        
    }];
}

- (void)sendEventToFlutterWith:(NSString *)key arguments:(NSDictionary *)arguments  completion:(void(^)(NSError* _Nullable))completion {
    CommonParams* params = [[CommonParams alloc] init];
    params.key = key;
    params.arguments = arguments;
    
    FlutterBasicMessageChannel *channel =
        [FlutterBasicMessageChannel
          messageChannelWithName:@"flutter.sendEvent.listener"
          binaryMessenger:self.binaryMessenger];
    [channel sendMessage:[params toMap] reply:^(id  _Nullable reply) {
        completion(nil);
    }];
}

- (void)setNativeMessageHandler {
    if (!_nativeChannel) {
        _nativeChannel =
            [FlutterBasicMessageChannel
              messageChannelWithName:@"flutter.sendEvent.native"
              binaryMessenger:self.binaryMessenger];
        
        __weak typeof(self) weakSelf = self;
        
        [_nativeChannel setMessageHandler:^(id _Nullable message, FlutterReply _Nonnull callback) {
            NSDictionary *dic = message;
            
            if (dic != nil) {
                
                __strong typeof(self) strongSelf = weakSelf;
                
                if (strongSelf.delegate == nil) {
                    callback(@{
                        @"error": @{
                            @"code": @"-1003",
                            @"message": @"delegate 未设置",
                            @"details": @"delegate 未设置，请使用setupDelegate初始化代理方法"
                        }
                    });
                    return;
                }
                
                CallMethodParams *params = [CallMethodParams fromMap:dic];
                
                if (params.methodName == nil || params.methodName.length == 0) {
                    callback(@{
                        @"error": @{
                            @"code": @"-1002",
                            @"message": @"methodName 不可为空",
                            @"details": @"methodName 不可为空!!!"
                        }
                    });
                    return;
                }
                
                NSString *methodName = params.methodName;
                
                if (strongSelf.delegate && [strongSelf.delegate respondsToSelector:@selector(callHandler:params:callback:)]) {
                    
                    [strongSelf.delegate callHandler:methodName params:params.arguments callback:^(id value) {
                        if ([value isKindOfClass:[NSDictionary class]]) {
                            [strongSelf sendCallbackToFlutter:methodName arguments:value];
                        } else {
                            // value的格式不正确
                        }
                    }];
                }
                
                callback(@{});
            }
            
            callback(@{
                @"error": @{
                    @"code": @"-1001",
                    @"message": @"message 格式不正确",
                    @"details": @"message 格式不正确!!!"
                }
            });
        }];
    }
}

- (void)sendCallbackToFlutter:(NSString *)methodName arguments:(NSDictionary*)arguments {
    ResponseParams* params = [[ResponseParams alloc] init];
    params.methodName = methodName;
    params.arguments = arguments;
    
    [self.nativeChannel sendMessage:[params toMap] reply:^(id  _Nullable reply) {
        
    }];
}

@end
